File: CellMesh/cellmesh/src/main/java/com/cellmesh/app/model/Node.java
package com.cellmesh.app.model;

import android.annotation.TargetApi;
import android.os.Build;
import android.util.Log;
import android.util.Xml;

import org.slf4j.impl.StaticLoggerBinder;
import org.w3c.dom.NodeList;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.acl.LastOwnerException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.Queue;

import io.netty.util.AsciiString;
import io.underdark.Underdark;
import com.cellmesh.app.MainActivity;
import io.underdark.transport.Link;
import io.underdark.transport.Transport;
import io.underdark.transport.TransportKind;
import io.underdark.transport.TransportListener;
import io.underdark.util.nslogger.NSLogger;
import io.underdark.util.nslogger.NSLoggerAdapter;

public class Node implements TransportListener
{
	private boolean running;
	private MainActivity activity;
	private long nodeId;
	private Transport transport;
	private INodeListener listener;
	private ArrayList<Link> links = new ArrayList<>();
	private Set<Long> ids=new HashSet<>();
	private String name;
	private nameManager nm;

	public Node(MainActivity activity,INodeListener listener,String name)
	{
		this.name=name;
		this.activity = activity;
		this.listener=listener;
		do
		{
			nodeId = new Random().nextLong();
		} while (nodeId == 0);

		if(nodeId < 0)
			nodeId = -nodeId;

		nm = new nameManager(nodeId, name);

		ids.add(nodeId);
		configureLogging();
		EnumSet<TransportKind> kinds = EnumSet.of(TransportKind.BLUETOOTH, TransportKind.WIFI);
		this.transport = Underdark.configureTransport(
				234235,
				nodeId,
				this,
				null,
				activity.getApplicationContext(),
				kinds
		);
	}

	public Map<Long,String> getNamesMap() {
		return Collections.unmodifiableMap(nm.getMap());
	}
	
	private void configureLogging()
	{
		NSLoggerAdapter adapter = (NSLoggerAdapter)
				StaticLoggerBinder.getSingleton().getLoggerFactory().getLogger(Node.class.getName());
		adapter.logger = new NSLogger(activity.getApplicationContext());
		adapter.logger.connect("192.168.5.203", 50000);
		Underdark.configureLogging(true);
	}

	public void start()
	{
		if(running)
			return;
		Log.d("Mesh","Starting Node");
		running = true;
		transport.start();
	}

	public void stop()
	{
		if(!running)
			return;

		Log.d("Mesh","Stopping Node");
		running = false;
		transport.stop();
	}

	public void broadcastMessage(String frameData)
	{
		if(links.isEmpty())
			return;
		listener.onDataSent(frameData,nodeId);
		for(Link link : links)
			link.sendFrame(frameData.getBytes());
	}

	public void sendMessage(String frameData, Link target)
	{
		target.sendFrame(frameData.getBytes());
	}

	//Call this when the names need to be updated by the UID
	private void doNameUpdate(){
		listener.onNamesUpdated(null);
	}
	private void handleEmergencyMessage(){
		listener.onEmergency(null);
	}
	//region TransportListener
	@Override
	public void transportNeedsActivity(Transport transport, ActivityCallback callback)
	{
		callback.accept(activity);
	}

	@Override
	public void transportLinkConnected(Transport transport, Link link)
	{
		Log.d("Mesh","Link "+Long.toString(link.getNodeId())+" Joined");
		links.add(link);
		ids.add(link.getNodeId());

		// Send our name data hash.
		sendMessage('1' + nm.getHash(), link);

		listener.onConnected(Collections.unmodifiableSet(ids),link.getNodeId());
	}

	@Override
	public void transportLinkDisconnected(Transport transport, Link link)
	{
		Log.d("Mesh","Link "+Long.toString(link.getNodeId())+" Left");
		ids.remove(link.getNodeId());
		links.remove(link);
		listener.onDisconnected(Collections.unmodifiableSet(ids),link.getNodeId());
	}

	private void compareHash(String hash, Link link) {
		if ( !hash.equals(nm.getHash()) ) {
			// Send our names.
			Map<Long, String> map = nm.getMap();
			for ( Map.Entry<Long,String> item : map.entrySet() ) {
				String message =  '2' + Long.toString(item.getKey()) + ':' + item.getValue();
				sendMessage(message, link);
			}
		}
	}
	@TargetApi(Build.VERSION_CODES.KITKAT)
	@Override
	public void transportLinkDidReceiveFrame(Transport transport, Link link, byte[] frameData)
	{
		/*
		00: reserved
		01: node hash received
		02: node data received
		10: chat message received
		 */
		int op = Integer.parseInt(new String(frameData, 0, 1, StandardCharsets.US_ASCII));
		String data = new String(frameData, 1, frameData.length, StandardCharsets.US_ASCII);

		switch(op) {
			case 0:
				// Reserved
				break;
			case 1:
				// Name hash received
				compareHash(data, link);
				break;
			case 2:
				// Name data received
				if ( data.indexOf(':') > 0 ) {
					Long srcId = Long.parseLong(data.substring(0, data.indexOf(':')));
					String name = data.substring(data.indexOf(':'), data.length());

					nm.addName(srcId, name);
				}
			case 10:
				// Message received
				listener.onDataReceived(data, link.getNodeId());
		}
	}
	//endregion
} // Node

File: app/src/main/java/com/cellmesh/nearby/nearbyapp/Node.kt
package com.cellmesh.nearby.nearbyapp

import android.support.v4.content.ContextCompat
import android.util.Log
import io.underdark.Underdark
import io.underdark.transport.Link
import io.underdark.transport.Transport
import io.underdark.transport.TransportKind
import io.underdark.transport.TransportKind.*
import io.underdark.transport.TransportListener
import java.util.*
import kotlin.collections.ArrayList
import io.underdark.util.nslogger.NSLogger
import io.underdark.util.nslogger.NSLoggerAdapter




/**
 * Created by Benjamin on 11/4/2017.
 */
class Node(
        val activity: MainActivity,
        val transport_types: EnumSet<TransportKind>,
        val nodeId: Long) : TransportListener {

    val links = mutableListOf<Link>()
    val transport = Underdark.configureTransport(
            234235,
            nodeId,
            this,
            null,
            activity.getApplicationContext(),
            transport_types)

    fun start(){
        val adapter = StaticLoggerBinder.getSingleton().getLoggerFactory().getLogger(Node::class.java.name) as NSLoggerAdapter
        adapter.logger = NSLogger(activity.applicationContext)
        adapter.logger.connect("192.168.5.203", 50000)
        Underdark.configureLogging(true)
        Log.d("Mesh","Starting transport")
        transport.start()
    }


    fun stop(){
        Log.d("Mesh","Stopping Transport")
        transport.stop()
    }
    fun broadcastBytes(data:ByteArray){
        Log.d("Mesh","Sending ${data.size} Bytes")
        for(link in links){
            link.sendFrame(data)
        }
    }
    fun sendBytes(data:ByteArray,nodeId: Long){
        links.
                firstOrNull { link->link.nodeId==nodeId }
                ?.sendFrame(data)
    }


    override fun transportNeedsActivity(transport: Transport?, callback: TransportListener.ActivityCallback?) {
        callback?.let { it.accept(activity) }
    }

    override fun transportLinkDidReceiveFrame(transport: Transport?, link: Link?, data: ByteArray?) {
        Log.d("Mesh","Got ${data?.let{it.size}} bytes from ${link?.let { it.nodeId }}")
    }

    override fun transportLinkConnected(transport: Transport?, link: Link?) {
        Log.d("Mesh","${link?.let{it.nodeId}} Connected")
        link?.let { links.add(it) }

        activity.handle_swarm_update(links.map { it.nodeId }.toList())
    }

    override fun transportLinkDisconnected(transport: Transport?, link: Link?) {
        Log.d("Mesh","${link?.let{it.nodeId}} Connected")
        link?.let { links.remove(it) }
        activity.handle_swarm_update(links.map { it.nodeId }.toList())
    }


}